flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/table/KuduCatalog.java

```diff
@@ -0,0 +1,348 @@
+import org.apache.kudu.client.KuduException;
+import org.apache.kudu.client.KuduTable;
+import org.apache.kudu.shaded.com.google.common.collect.Sets;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.connectors.kudu.table.KuduTableFactory.KUDU_HASH_COLS;
+import static org.apache.flink.connectors.kudu.table.KuduTableFactory.KUDU_MASTERS;
+import static org.apache.flink.connectors.kudu.table.KuduTableFactory.KUDU_PRIMARY_KEY_COLS;
+import static org.apache.flink.connectors.kudu.table.KuduTableFactory.KUDU_REPLICAS;
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Catalog for reading and creating Kudu tables.
+ */
+@PublicEvolving
+public class KuduCatalog extends AbstractReadOnlyCatalog {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KuduCatalog.class);
+    private final KuduTableFactory tableFactory = new KuduTableFactory();
+    private final String kuduMasters;
+    private KuduClient kuduClient;
+
+    /**
+     * Create a new {@link KuduCatalog} with the specified catalog name and kudu master addresses.
+     *
+     * @param catalogName Name of the catalog (used by the table environment)
+     * @param kuduMasters Connection address to Kudu
+     */
+    public KuduCatalog(String catalogName, String kuduMasters) {
+        super(catalogName, EnvironmentSettings.DEFAULT_BUILTIN_DATABASE);
+        this.kuduMasters = kuduMasters;
+        this.kuduClient = createClient();
+    }
+
+    /**
+     * Create a new {@link KuduCatalog} with the specified kudu master addresses.
+     *
+     * @param kuduMasters Connection address to Kudu
+     */
+    public KuduCatalog(String kuduMasters) {
+        this("kudu", kuduMasters);
+    }
+
+    public Optional<TableFactory> getTableFactory() {
+        return Optional.of(getKuduTableFactory());
+    }
+
+    public KuduTableFactory getKuduTableFactory() {
+        return tableFactory;
+    }
+
+    private KuduClient createClient() {
+        return new KuduClient.KuduClientBuilder(kuduMasters).build();
+    }
+
+    @Override
+    public void open() {}
+
+    @Override
+    public void close() {
+        try {
+            if (kuduClient != null) {
+                kuduClient.close();
+            }
+        } catch (KuduException e) {
+            LOG.error("Error while closing kudu client", e);
+        }
+    }
+
+    public ObjectPath getObjectPath(String tableName) {
+        return new ObjectPath(getDefaultDatabase(), tableName);
+    }
+
+    @Override
+    public List<String> listTables(String databaseName) throws DatabaseNotExistException, CatalogException {
```

Comment:
This api is a bit unusual given the call to kudu isn't actually filtering by "database" at all. 

Kudu doesn't have a strict first-class database concept yet, though it does have a table name syntax to help loosely represent a database. The format is effectively `<database>.<tablename>`. 

Handling this is likely out of scope for this patch, but I wanted to provide some context.

Reply:
For now we decided to not introduce any database concept in this catalog for the reasons you highlighted and treat everything as a single database.

In the future we could improve this and adapt the convention you mentioned. In that case the default could be the global namespace as it is now.